src/commands/registerCommands.ts

```diff
@@ -179,24 +181,83 @@ async function grammarAssociationCommand(documentURI: Uri, languageClient: Langu
   const fileUri = await window.showOpenDialog(options);
   if (fileUri && fileUri[0]) {
     // The XSD, DTD has been selected, get the proper syntax for binding this grammar file in the XML document.
-    const identifier = TextDocumentIdentifier.create(documentURI.toString());
     const grammarURI = fileUri[0];
     try {
-      const result = await commands.executeCommand(ServerCommandConstants.ASSOCIATE_GRAMMAR_INSERT, identifier, grammarURI.toString(), bindingType);
-      // Insert the proper syntax for binding
-      const lspTextDocumentEdit = <TextDocumentEdit>result;
-      const workEdits = new WorkspaceEdit();
-      for (const edit of lspTextDocumentEdit.edits) {
-        workEdits.replace(documentURI, languageClient.protocol2CodeConverter.asRange(edit.range), edit.newText);
+      if (bindingType == 'fileAssociation') {
+        // Bind grammar using file association
+        await bindWithFileAssociation(documentURI, grammarURI);
+      } else {
+        // Bind grammar using standard binding
+        await bindWithStandard(documentURI, grammarURI, bindingType, languageClient);
       }
-      workspace.applyEdit(workEdits); // apply the edits
-
     } catch (error) {
       window.showErrorMessage('Error during grammar binding: ' + error.message);
     };
   }
 }
 
+/**
+ * Perform grammar binding using file association through settings.json
+ *
+ * @param documentURI the URI of the current XML document
+ * @param grammarURI the URI of the user selected grammar file
+ */
+async function bindWithFileAssociation(documentURI: Uri, grammarURI: Uri) {
+  const grammarFilePath = grammarURI.toString();
+  const currentFilename = path.basename(documentURI.toString());
+  const currentWorkspaceUri = getAbsoluteCurrentWorkspaceUri();
+  const relativeGrammarFilePath = getRelativeUri(currentWorkspaceUri, grammarFilePath);
```

Comment:
If user select a xsd which is not inside workspace it should return the uri whith file:// is it working like this?


I will rename relatibegrammarfilepath with grammarFilePath singe the path can be relative to workspace or not

Reply:
From https://github.com/redhat-developer/vscode-xml/pull/584#discussion_r690194431, at the moment the implemented behaviour is ii (the relative path). Did you want i (the absolute path) instead?

Reply:
in your example for ii, folder catalog-preferences inside the workspace?

Reply:
And would it be sufficient to just use the relative path, or would you like the i and ii cases?

Reply:
I just made a fix where the grammarFilePath now depends on if the file is in/out of the workspace folder